### PR TITLE
VReplication: Make Target Sequence Initialization More Robust

### PR DESCRIPTION
**Pull request by mattlord**
## Description

This makes two key improvements to how we manage the initialization of target [sequences](https://vitess.io/docs/reference/features/vitess-sequences/) when executing a [`MoveTables SwitchTraffic`](https://vitess.io/docs/reference/vreplication/movetables/#switchtraffic) command and the [`--initialize-target-sequences`](https://vitess.io/docs/reference/programs/vtctldclient/vtctldclient_movetables/vtctldclient_movetables_switchtraffic/#options) flag is specified:
1. We properly account for the case where the target table using the sequence is empty on a target shard
2. We treat these failures as rollback / revert eligible. This makes sense because if we have not successfully completed the sequence init work then we cannot guarantee that there will not be insert errors when using the sequence after we switch traffic AND reverting back is safe AND we can safely attempt to reinitialize the target sequences again on the next attempt (assuming the failure was ephemeral or is otherwise addressed)

I think that this fix should be backported to v18 — where the initial feature was first added — as the changes are isolated to users of the new feature and this work makes that feature more robust and reliable for them.

<details>
  <summary>Click here to see a manual test</summary>

```
pushd ${VTROOT}/examples/local

./101_initial_cluster.sh

mysql < ../common/insert_commerce_data.sql

mysql commerce -e "insert into customer (customer_id, email) values (99999, 'mlord@planetscale.com')" --binary-as-hex=false

vtctldclient --server localhost:15999 CreateKeyspace --durability-policy=semi_sync customer || fail "Failed to create and configure the commerce keyspace"

./301_customer_sharded.sh; ./302_new_shards.sh

sleep 30

vtctldclient ApplySchema --sql-file create_customer_sharded.sql customer

mysql commerce -e "select * from customer order by customer_id" --binary-as-hex=false
mysql commerce -e "select * from customer_seq" --binary-as-hex=false

vtctldclient MoveTables --target-keyspace customer --workflow commerce2customer create --tables 'customer,corder' --source-keyspace commerce

sleep 30

mysql customer -e "select * from customer order by customer_id" --binary-as-hex=false

# Delete the record that is on the 80- shard
primaryuid=$(vtctldclient GetTablets --keyspace customer --tablet-type primary --shard "80-" | awk '{print $1}' | cut -d- -f2 | bc)
command mysql -u root --socket=${VTDATAROOT}/vt_0000000${primaryuid}/mysql.sock --binary-as-hex=false vt_customer -e "delete from vt_customer.customer"

vtctldclient MoveTables --target-keyspace customer --workflow commerce2customer SwitchTraffic --initialize-target-sequences

mysql commerce -e "select * from customer_seq" --binary-as-hex=false

mysql customer -e "insert into customer (email) values ('mlord2@planetscale.com')" --binary-as-hex=false

mysql commerce -e "select * from customer_seq" --binary-as-hex=false

sleep 5

mysql customer -e "select * from customer order by customer_id" --binary-as-hex=false
```
</details>

## Related Issue(s)

  - Fixes: https://github.com/vitessio/vitess/issues/15295
  - Follow-up to: https://github.com/vitessio/vitess/pull/13656

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtctl/workflow/server.go [hunk 1]
```diff
@@ -3301,6 +3301,20 @@ func (s *Server) switchWrites(ctx context.Context, req *vtctldatapb.WorkflowSwit
 			sw.cancelMigration(ctx, sm)
 			return handleError("failed to create the reverse vreplication streams", err)
 		}
+
+		// Initialize any target sequences, if there are any, before allowing new writes.
+		if req.InitializeTargetSequences && len(sequenceMetadata) > 0 {
+			ts.Logger().Infof("Initializing target sequences")
+			// Writes are blocked so we can safely initialize the sequence tables but
+			// we also want to use a shorter timeout than the parent context.
+			// We use up at most half of the overall timeout.
```
Comment (ajm188) [inline go/vt/vtctl/workflow/server.go:3310]:
```suggestion
			// We use at most half of the overall timeout.
```

File: go/vt/vtctl/workflow/traffic_switcher.go [hunk 2]
```diff
@@ -1432,13 +1432,17 @@ func (ts *trafficSwitcher) initializeTargetSequences(ctx context.Context, sequen
 				MaxRows: 1,
 			})
 			if terr != nil || len(qr.Rows) != 1 {
-				return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to get the max used sequence value for target table %s.%s in order to initialize the backing sequence table: %v",
-					ts.targetKeyspace, sequenceMetadata.usingTableName, terr)
+				return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to get the max used sequence value for target table %s.%s on tablet %v in order to initialize the backing sequence table: %v",
+					ts.targetKeyspace, sequenceMetadata.usingTableName, primary.Alias, terr)
```
Comment (ajm188) [inline go/vt/vtctl/workflow/traffic_switcher.go:1436]:
```suggestion
					ts.targetKeyspace, sequenceMetadata.usingTableName, topoproto.TabletAliasString(primary.Alias), terr)
```

File: go/vt/wrangler/traffic_switcher.go [hunk 3]
```diff
@@ -619,6 +619,20 @@ func (wr *Wrangler) SwitchWrites(ctx context.Context, targetKeyspace, workflowNa
 			sw.cancelMigration(ctx, sm)
 			return handleError("failed to create the reverse vreplication streams", err)
 		}
+
+		// Initialize any target sequences, if there are any, before allowing new writes.
+		if initializeTargetSequences && len(sequenceMetadata) > 0 {
+			ts.Logger().Infof("Initializing target sequences")
+			// Writes are blocked so we can safely initialize the sequence tables but
+			// we also want to use a shorter timeout than the parent context.
+			// We use up at most half of the overall timeout.
```
Comment (ajm188) [inline go/vt/wrangler/traffic_switcher.go:628]:
```suggestion
			// We use at most half of the overall timeout.
```

File: go/vt/wrangler/traffic_switcher.go [hunk 4]
```diff
@@ -2197,13 +2200,17 @@ func (ts *trafficSwitcher) initializeTargetSequences(ctx context.Context, sequen
 			)
 			qr, terr := ts.wr.ExecuteFetchAsApp(ictx, primary.GetAlias(), true, query.Query, 1)
 			if terr != nil || len(qr.Rows) != 1 {
-				return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to get the max used sequence value for target table %s.%s in order to initialize the backing sequence table: %v",
-					ts.targetKeyspace, sequenceMetadata.usingTableName, terr)
+				return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to get the max used sequence value for target table %s.%s on tablet %v in order to initialize the backing sequence table: %v",
+					ts.targetKeyspace, sequenceMetadata.usingTableName, primary.Alias, terr)
```
Comment (ajm188) [inline go/vt/wrangler/traffic_switcher.go:2204]:
```suggestion
					ts.targetKeyspace, sequenceMetadata.usingTableName, topoproto.TabletAliasString(primary.Alias), terr)
```

File: go/vt/wrangler/traffic_switcher.go [hunk 5]
```diff
@@ -2197,13 +2200,17 @@ func (ts *trafficSwitcher) initializeTargetSequences(ctx context.Context, sequen
...
-			if terr != nil {
-				return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to get the max used sequence value for target table %s.%s in order to initialize the backing sequence table: %v",
-					ts.targetKeyspace, sequenceMetadata.usingTableName, terr)
+			rawVal := sqltypes.Proto3ToResult(qr).Rows[0][0]
+			maxID := int64(0)
+			if !rawVal.IsNull() { // If it's NULL then there are no rows and 0 remains the max
+				maxID, terr = rawVal.ToInt64()
+				if terr != nil {
+					return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to get the max used sequence value for target table %s.%s on tablet %v in order to initialize the backing sequence table: %v",
+						ts.targetKeyspace, sequenceMetadata.usingTableName, primary.Alias, terr)
```
Comment (ajm188) [inline go/vt/wrangler/traffic_switcher.go:2212]:
```suggestion
						ts.targetKeyspace, sequenceMetadata.usingTableName, topoprot.TabletAliasString(primary.Alias), terr)
```

File: go/vt/vtctl/workflow/traffic_switcher.go [hunk 6]
```diff
@@ -1432,13 +1432,17 @@ func (ts *trafficSwitcher) initializeTargetSequences(ctx context.Context, sequen
...
-			if terr != nil {
-				return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to get the max used sequence value for target table %s.%s in order to initialize the backing sequence table: %v",
-					ts.targetKeyspace, sequenceMetadata.usingTableName, terr)
+			rawVal := sqltypes.Proto3ToResult(qr).Rows[0][0]
+			maxID := int64(0)
+			if !rawVal.IsNull() { // If it's NULL then there are no rows and 0 remains the max
+				maxID, terr = rawVal.ToInt64()
+				if terr != nil {
+					return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to get the max used sequence value for target table %s.%s on tablet %v in order to initialize the backing sequence table: %v",
+						ts.targetKeyspace, sequenceMetadata.usingTableName, primary.Alias, terr)
```
Comment (ajm188) [inline go/vt/vtctl/workflow/traffic_switcher.go:1444]:
```suggestion
						ts.targetKeyspace, sequenceMetadata.usingTableName, topoproto.TabletAliasString(primary.Alias), terr)
```

**Review by ajm188 — APPROVED** [review]
Overall LGTM. Some suggestions around the log messages purely for readability. I agree with the call to backport to 18.0

File: go/vt/vtctl/workflow/traffic_switcher.go (hunk 2 above)
Comment (mattlord) [reply go/vt/vtctl/workflow/traffic_switcher.go:1436]: I thought about using that as well (so it matches the format e.g. in `GetTablets`) and went back and forth on it. If others also prefer that then I'll switch. Will plan on it.

**Review by rohit-nayak-ps — APPROVED** [review]

**Review by ajm188 — APPROVED** [review]

**Pull request merged**